nemo/collections/nlp/models/token_classification/token_classification_model.py

```diff
@@ -43,16 +47,28 @@ def input_types(self) -> Optional[Dict[str, NeuralType]]:
     def output_types(self) -> Optional[Dict[str, NeuralType]]:
         return self.classifier.output_types
 
+    @classmethod
+    def update_config_with_specific_artifacts(cls, config: OmegaConf, artifacts_dir: str = None) -> OmegaConf:
+        config.data_desc_pickle = os.path.join(artifacts_dir, config.data_desc_pickle)
+        return config
+
     def __init__(self, cfg: DictConfig, trainer: Trainer = None):
-        """
-        Initializes BERT Named Entity Recognition model.
-        """
+        """Initializes Token Classification Model."""
 
-        self.data_desc = TokenClassificationDataDesc(
-            data_dir=cfg.dataset.data_dir, modes=["train", "test", "dev"], pad_label=cfg.dataset.pad_label
-        )
         self.data_dir = cfg.dataset.data_dir
 
+        modes = ["train", "test", "dev"]
+        if not os.path.exists(cfg.data_desc_pickle):
+            # process the data and extract label_ids and class weights if the model is not restored from .nemo file
+            self.data_desc = TokenClassificationDataDesc(
+                data_dir=self.data_dir, modes=modes, pad_label=cfg.dataset.pad_label
+            )
+            if not self.data_desc.data_found:
+                raise ValueError(f'No {modes} data found at {self.data_dir}.')
+        else:
+            self.data_desc = pickle.load(open(cfg.data_desc_pickle, "rb"))
```

Comment:
Use register artifact function to properly return a correct path rather than do manually checks

Reply:
In the config I provide only the name of the data_desc_pickle, I create the self.data_desc during data preprocessing step that is done during the model init, so I have to store self.data_desc first

Reply:
Ok thats another issue then, the setup_* methods can be called multiple times (either if it supports multiple data loaders which I think this script not support thankfully, and during .restore_from()). I see that the file being saved is not marked with unique name, so basically if it gets called multiple time, previous values will get overwritten. Is it possible to refactor this outside of the setup method, or to provide a unique name inside it?

Reply:
I'm saving this file inside setup training data only, why would it be called multiple times? Didn't you disable setup_data calls for restore_from()?

Reply:
Setup disable calls are attempted by the restore from step (since it's in the constructor). It will fail on other machines and print a log message since paths won't match, and so I guess if won't save file either. That's fine then.